### DISCUSSION TITLE
Adding mulitple variables at once in OPCUA Server

### DISCUSSION THREAD

**Discussion by vidyarao123** [FreeOpcUa/python-opcua#1257] [Q&A]
Hi,

We are working on adding  approximately 20000 variables in OPCUA Server from CSV / Excel. The addition is successful but iterating the below line by using a for loop for 20000 variables is a time consuming task. 
**OPC_Tag = Param.add_variable(addspace,Tag,0)**

So we wanted to know, if there is a way to add such huge quantity of tags at a faster rate to server. 

Thanks

**Comment by AndreasHeine** [comment]
does they change while runtime or are they always the same?

**Reply by vidyarao123** [reply]
Its one time

**Reply by vidyarao123** [reply]
Once loaded, they wont change during the runtime

**Reply by AndreasHeine** [reply]
if i remember right there is the possibility to use a shelffile!

**Reply by vidyarao123** [reply]
Do you mean this module ?
shelve — Python object persistence¶

**Reply by vidyarao123** [reply]
Correct me if I am wrong. The shelve module is just an alternative for the list as I understand. But I am trying to achieve is the add all variables parallely. Instead of a sequential method.

**Reply by AndreasHeine** [reply]
you should look in the code below

Server:
https://github.com/FreeOpcUa/python-opcua/blob/0d595e596c9922699f0b36abb79583d5763eeb9b/opcua/server/server.py#L80
https://github.com/FreeOpcUa/python-opcua/blob/0d595e596c9922699f0b36abb79583d5763eeb9b/opcua/server/server.py#L92

InternalServer:
https://github.com/FreeOpcUa/python-opcua/blob/0d595e596c9922699f0b36abb79583d5763eeb9b/opcua/server/internal_server.py#L100
https://github.com/FreeOpcUa/python-opcua/blob/0d595e596c9922699f0b36abb79583d5763eeb9b/opcua/server/internal_server.py#L159
https://github.com/FreeOpcUa/python-opcua/blob/0d595e596c9922699f0b36abb79583d5763eeb9b/opcua/server/internal_server.py#L165

**Reply by zerox1212** [reply]
> if i remember right there is the possibility to use a shelffile!

Shelf only works for standard address space! It won't save user added nodes!

**Comment by oroulet** [comment]
look at code of create_variable() and write your low level code yourself.
But anyway adding new nodes is crazy time consuming. That is why we do not read the xml standard address space from xml at startup but generate python code from the xml and execute it.

**Comment by zerox1212** [comment]
I have used `dump` and `load` to make the nodes in the address space persistent. Then when server starts just check if the dump file exists, and if it does load from there. Server will start very fast.

Note that this is different than using a shelf!!!
https://github.com/FreeOpcUa/python-opcua/blob/0d595e596c9922699f0b36abb79583d5763eeb9b/opcua/server/address_space.py#L563

Read the docs around those functions though... they have some limitations.
